### DISCUSSION TITLE
YOLO Inferência Thread-Safe - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#7701] [Docs]
# YOLO Inferência Thread-Safe - Ultralytics YOLOv8 Docs

Este guia fornece as melhores práticas para a realização de inferência thread-safe com modelos YOLO , garantindo previsões fiáveis e simultâneas em aplicações multi-threaded.

https://docs.ultralytics.com/pt/guides/yolo-thread-safe-inference/

**Comment by WellingtonDev25** [comment]
Excellent! Do you have a similar explanation about YOLO inference in multiprocessing? Seems to generate memory errors especially using CUDA (GPU)

**Reply by pderrenger** [reply]
@WellingtonDev25 when using YOLO models with multiprocessing, especially on CUDA, it's important to ensure that each process manages its own GPU memory to prevent conflicts. Here are some tips:

1. **Instantiate the Model in Each Process**: Create a new model instance within each subprocess rather than sharing a model across processes.
2. **Manage CUDA Contexts**: Each process should have its own CUDA context. This is usually handled automatically if you're using PyTorch's `multiprocessing` module.
3. **Control GPU Usage**: Set the `CUDA_VISIBLE_DEVICES` environment variable within each subprocess to control which GPU is used.
4. **Conserve Memory**: Use the `--device` argument to specify the GPU, and consider using the `--imgsz` argument to adjust the input size for memory efficiency.
5. **Clean Up**: Ensure that each process frees up its GPU memory properly after use by calling `torch.cuda.empty_cache()`.

For more detailed guidance, please refer to our documentation on [Predict Mode](https://docs.ultralytics.com/modes/predict) and [Best Practices](https://docs.ultralytics.com/guides/best-practices). Remember to handle the unique GPU memory management requirements when using multiprocessing to avoid memory errors. 🚀

**Reply by WellingtonDev25** [reply]
Great! thanks @pderrenger

**Reply by pderrenger** [reply]
@WellingtonDev25 you're welcome! If you have any more questions or need further assistance, feel free to ask. Happy coding with YOLOv8! 😊👍
